### ISSUE TITLE
gradient for collective apis

### ISSUE THREAD

**Issue by ailzhang** [pytorch/xla#2989]
Reported by @ronghanghu in https://github.com/pytorch/xla/pull/2086#issuecomment-857394739.

**Comment by ronghanghu** [comment]
Following up on this issue: I believe one can copy of the implementation from PyTorch's `torch.distributed.nn.*` in https://github.com/pytorch/pytorch/blob/8c798e062216278673a75bac0848ea69a8bd3f03/torch/distributed/nn/functional.py#L210-L263 for autograd version of all_reduce and all_gather.

**Comment by JackCaoG** [comment]
@ronghanghu I saw https://github.com/pytorch/pytorch/issues/58005 claims `torch.distributed.nn.*` implementation incorrectly scale up the gradients and it will be fixed in the upcoming releases. Could you confirm that this is the expected behavior? It is kind of confusing that `torch.distributed.nn.all_reduce` and `torch.distributed.all_reduce` has different behavior.

**Comment by ronghanghu** [comment]
@JackCaoG Yes, I believe the current `torch.distributed.nn.*` implementation is the desired behavior. I replied in https://github.com/pytorch/pytorch/issues/58005#issuecomment-870214108, but perhaps I can wait for a confirmation from the PyTorch maintainer in https://github.com/pytorch/pytorch/issues/58005

**Comment by JackCaoG** [comment]
@ronghanghu thanks! It would be nice if we can get some confirmation from the pytorch maintainer that this behavior will persist.
